module/localisation/RobotParticleLocalisation/src/RobotModel.h

```diff
@@ -47,42 +62,183 @@ namespace localisation {
+
+            // rZFf = vector from field origin to zenith high in the sky
+            Eigen::Vector4d rZFf = {0, 0, 1, 0};
+            rZFf                 = Hcf * rZFf;
+
+            if (type == Goal::MeasurementType::CENTRE) {
+                // rGCc = vector from camera to goal post expected position
+                Eigen::Vector4d rGCc_4   = {actual_position[0], actual_position[1], actual_position[2], 1};
+                rGCc_4                   = Hcf * rGCc_4;
+                Eigen::Vector3d rGCc     = {rGCc_4[0], rGCc_4[1], rGCc_4[2]};
+                Eigen::Vector3d rGCc_sph = cartesianToSpherical(rGCc);  // in r,theta,phi
+                return rGCc_sph;
+            }
+
+            switch (FieldDescription::GoalpostType::Value(fd.dimensions.goalpost_type)) {
+                case FieldDescription::GoalpostType::CIRCLE: {
+                    if (type == Goal::MeasurementType::LEFT_NORMAL || type == Goal::MeasurementType::RIGHT_NORMAL) {
+                        Eigen::Vector3d rNCc   = getCylindricalPostCamSpaceNormal(type, actual_position, Hcf, fd);
+                        Eigen::Vector2d angles = {
+                            std::atan2(rNCc[1], rNCc[0]),
+                            std::atan2(rNCc[2], std::sqrt(rNCc[0] * rNCc[0] + rNCc[1] * rNCc[1]))};
+                        return angles;
+                    }
+                    break;
+                }
+                case FieldDescription::GoalpostType::RECTANGLE: {
+                    if (type == Goal::MeasurementType::LEFT_NORMAL || type == Goal::MeasurementType::RIGHT_NORMAL) {
+                        Eigen::Vector3d rNCc   = getSquarePostCamSpaceNormal(type, actual_position, Hcf, fd);
+                        Eigen::Vector2d angles = {
+                            std::atan2(rNCc[1], rNCc[0]),
+                            std::atan2(rNCc[2], std::sqrt(rNCc[0] * rNCc[0] + rNCc[1] * rNCc[1]))};
+                        return angles;
+                    }
+                    break;
+                }
+            }
+            return Eigen::Vector2d({0, 0});
+        }
 
-        arma::vec observationDifference(const arma::vec& a, const arma::vec& b);
+        StateVec limit(const StateVec& state) {
+            auto state2 = state;
+            // state2[kAngle] = normalizeAngle(state2[kAngle]);
+            // TODO: Clip robot's state to the field?
+            return state2;
+        }
 
-        arma::vec::fixed<size> limitState(const arma::vec::fixed<size>& state);
+        StateMat noise(const Scalar& deltaT) {
+            return processNoiseDiagonal.asDiagonal() * deltaT;
+        }
 
-        arma::mat::fixed<size, size> processNoise();
+        template <typename... Args>
+        Eigen::Matrix<Scalar, 1, 1> predict(const StateVec& state, const Args&... params) {
 
-        arma::vec3 processNoiseDiagonal;
+            // Our prediction is the first state
+            return Eigen::Matrix<Scalar, 1, 1>(state[kX]);
+        }
```

Comment:
Remove this

Reply:
It doesn't build if I remove the function

Reply:
you will have to remove this and fix up `predictedObservation` at the same time